minard/pingcratesdb.py

```diff
@@ -1,65 +1,196 @@
+            # Then its probably a mismapping
+            mismapped_n100+=str(crate)+", "
+
+    # Similiarly for N20
+    for crate in removed_crate_n20:
+        if crate not in n20_failed:
+            mismapped_n20+=str(crate)+", "
+
+    # Formatting
+    n100_str = n100_str[0:-2]
+    n20_str = n20_str[0:-2]
+    mismapped_n100 = mismapped_n100[0:-2]
+    mismapped_n20 = mismapped_n20[0:-2]
+
+    return n100_str, n20_str, mismapped_n100, mismapped_n20
+
+
+def crates_failed_messages(run):
     '''
-    Returns Redis entries for specific run by run number
-    Requires Redis instance, and run number.
+    Returns messages related to ping crates failing 
+    as well as potential MTCA/+ relay mis-mappings
     '''
-    keys = redis.zrangebyscore(RUN_INDEX, runnum, runnum)
-    p = redis.pipeline()
-    for key in keys:
-        p.hgetall(key)
-    return p.execute()    
 
-def del_run_from_db(run_number):
+    messages = []
+    if run < PING_CRATES_START_RUN:
+        return messages
+
+    # Warn about ping crates failures, messages print to detector state check
+    try:
+        n100, n20, mn100, mn20 = crates_failed(run)
+        if n100:
+            if len(n100) > 2:
+                messages.append("crates %s failed N100 checks in ping crates" % n100)
+            else:
+                messages.append("crate %s failed N100 checks in ping crates" % n100)
+        if n20:
+            if len(n20) > 2:
+                messages.append("crates %s failed N20 checks in ping crates" % n20)
+            else:
+                messages.append("crate %s failed N20 checks in ping crates" % n20)
+        if mn100:
+            if len(mn100) > 2:
+                messages.append("Warning: crates %s N100 is mismapped !" % mn100)
+            else:
+                messages.append("Warning: crate %s N100 is mismapped !" % mn100)
+        if mn20:
+            if len(mn20) > 2:
+                messages.append("Warning: crates %s N20 is mismapped !" % mn20)
+            else:
+                messages.append("Warning: crate %s N20 is mismapped !" % mn20)
+    except Exception as e:
+        # No avaiable ping crates data, no need to warn
+        pass
+
+    return messages
+
+
+def ping_crates_list(limit):
     '''
-    Delete run from Redis. Requires Redis instance and run number. 
+    Returns a list of ping crates information for runs larger
+    than the current run - limit
     '''
-    key = 'pingcrates-run-%s' % run_number
-    p = redis.pipeline()
-    p.delete(key)
-    p.zrem(RUN_INDEX, key)
-    p.zrem(TIME_INDEX, key)
-    return p.execute()
- 
+
+    run = get_latest_run()
+
+    conn = engine.connect()
+
+    result = conn.execute("SELECT timestamp, run, n100_crates_failed, n20_crates_failed FROM "
+                          "(SELECT *, ROW_NUMBER() OVER (PARTITION BY run ORDER BY timestamp DESC) "
+                          "AS ROWNUM FROM ping_crates) x WHERE ROWNUM = 1 AND run > %s", \
+                          (run-limit))
```

Comment:
Is this supposed to get the latest result for each run? Could you do something like:

```
SELECT DISTINCT ON (run) timestamp, run, n100_crates_failed, n20_crates_failed FROM ping_crates WHERE run > %s ORDER BY run, timestamp DESC
```

Reply:
That would be much simpler .... I was so proud of my fancy select statement :octopus: